File: src/Functional/Clojure/adapter.clj
(defn legacy-read-fahrenheit [] 86)

(defn adapt-to-celsius [read-fahrenheit]
  (fn [] (quot (* (- (read-fahrenheit) 32) 5) 9)))

(let [read-celsius (adapt-to-celsius legacy-read-fahrenheit)]
  (println (str "legacy=" (legacy-read-fahrenheit) "F"))
  (println (str "adapted=" (read-celsius) "C")))

File: src/Functional/Elixir/adapter.exs
defmodule LegacyFahrenheitSensor do
  def read_fahrenheit, do: 86
end

defmodule FahrenheitSensorAdapter do
  def read_celsius(sensor_module) do
    fahrenheit = apply(sensor_module, :read_fahrenheit, [])
    div((fahrenheit - 32) * 5, 9)
  end
end

IO.puts("legacy=#{LegacyFahrenheitSensor.read_fahrenheit()}F")
IO.puts("adapted=#{FahrenheitSensorAdapter.read_celsius(LegacyFahrenheitSensor)}C")
